File: python/bbqr/split.py
#
# (c) Copyright 2023 by Coinkite Inc. This file is in the public domain.
#
# - uses pyqrcode for deep QR knowledge
# - text prefix on each QR:
#
#       B$                  fixed header for this protocol (2 chars)
#       2                   one char encode: Z=zlib, 2=Base32, H=Hex
#       P                   one char file type: P=PSBT, T=TXN, etc
#       05                  2-digits of base 36: number of data QR codes (all are needed)
#       00                  2-digits of base 36: which QR code this is in the sequence
#
# - optional parity QR codes use indexes 05 and up; any 5 of the series recover the data
#
from math import ceil
from .utils import version_to_chars, encode_bytes, bytes_to_text, pad_block, int2base36
from .consts import HEADER_LEN, KNOWN_FILETYPES, MAX_PARITY_INDEX
from .consts import ENCODING_BYTE_MOD, ENCODING_SPLIT_MOD
from . import gf256

def num_qr_needed(ver, ll, split_mod, reserve=0):
    # Determine number of QR's at indicated version would be
    # needed to hold ll characters. when 2 or more QR, consider
    # the exact split point cannot be between encoded symbols
    # - ok to return huge numbers for unlikely cases
    # - reserve: chars to leave free in each QR (parity parts are one group longer)
    cap = version_to_chars(ver) - HEADER_LEN - reserve
    cap2 = cap - (cap % split_mod)

    need = ceil(ll / cap2)

    if need == 1:
        # no alignment concerns
        assert ll <= cap
        return 1, ll

    # going to be 2 or more, gotta be precise
    actual = ((need - 1) * cap2) + cap

    return (need if actual >= ll else (need + 1)), cap2

def find_best_version(ll, split_mod, min_split=1, max_split=1295, min_version=5, max_version=40, reserve=0):
    # Find ideal QR version and provide # of QR and splits needed.
    # - assumes you want to pack the QR, so forcing min_split means you need to have the data
    #   at least the data to fill that # of QR at min_version
    #
    # ll = length of encoded data to be transmitted (no headers)
    # split_mod = required size of non-runt parts so that can be decoded w/o spliting symbols

    min_version = min(min_version, max_version)     # in case they spec a very low max

    assert 1 <= min_version <= max_version <= 40, "min/max version out of range"
    assert 1 <= min_split <= max_split <= 1295, "num splits out of range"

    options = []
    for ver in range(min_version, max_version+1):
        count, pe = num_qr_needed(ver, ll, split_mod, reserve)
        if not (min_split <= count <= max_split): continue
        options.append( (ver, count, pe) )

    # pick smallest number of QR, lowest version
    options.sort(key=lambda v: (v[1], v[0]))

    if not options:
        raise ValueError("Cannot make it fit")

    return options[0]

def parity_parts(stream, num_qr, per_each, parity, encoding):
    # Build `parity` extra parts (text, without headers) for a series of num_qr data parts.
    # - any num_qr parts out of the whole series recover the data
    # - each data block is padded with 0x80 then zeros to one symbol group more than a
    #   full block, so a rebuilt block always shows where its data ends
    # - Reed-Solomon over GF(2^8): block i is the value at x=i, see gf256.py
    assert num_qr >= 2, 'parity needs 2 or more data parts'
    assert num_qr + parity <= MAX_PARITY_INDEX + 1, 'too many parts for parity'

    mod = ENCODING_BYTE_MOD[encoding]
    split_mod = ENCODING_SPLIT_MOD[encoding]
    b_data = (per_each // split_mod) * mod            # bytes in a full data part
    b_rs = b_data + mod

    blocks = [pad_block(stream[off:off+b_data], b_rs) for off in range(0, len(stream), b_data)]
    assert len(blocks) == num_qr

    return [bytes_to_text(p, encoding) for p in gf256.encode_parity(blocks, num_qr + parity)]

def split_qrs(raw, type_code, encoding=None, parity=0, **kws):
    # Take some bytes and yield a series of text values that 
    # can be sent as QR code.
    # - returns text
    # - assumes and requires alnum, L error level
    # - parity: number of extra QR codes to add; any N of the whole series recover the data
    #   (no effect when everything fits into a single QR)
    # - kws: min_split, max_split, min_version, max_version (see find_best_version)

    assert type_code in KNOWN_FILETYPES, f"invalid type_code: {type_code}"
    if encoding: assert encoding in 'H2Z', f"invalid encoding: {encoding}"
    if not isinstance(raw, bytes):
        assert isinstance(raw, str), "need binary or text"
        raw = raw.encode('utf-8')
    assert isinstance(parity, int) and not isinstance(parity, bool) \
            and 0 <= parity < MAX_PARITY_INDEX, f"invalid parity: {parity!r}"

    # perhaps compress data
    encoding, stream = encode_bytes(raw, encoding)
    encoded = bytes_to_text(stream, encoding)
    split_mod = ENCODING_SPLIT_MOD[encoding]

    ll = len(encoded)

    ver, num_qr, per_each = find_best_version(ll, split_mod, **kws)

    if parity and num_qr >= 2:
        # parity parts are one symbol group longer than data parts; leave room for
        # that, now that we know the data needs more than one QR anyway
        ver, num_qr, per_each = find_best_version(ll, split_mod, reserve=split_mod, **kws)
    else:
        # a single QR has nothing to lose
        parity = 0

    assert per_each * num_qr >= ll

    hdr = f'B${encoding}{type_code}' + int2base36(num_qr)

    parts = [hdr + int2base36(n) + encoded[off:off+per_each] for
                            (n, off) in enumerate(range(0, ll, per_each))]

    if parity:
        parts += [hdr + int2base36(num_qr + n) + p for n, p in
                    enumerate(parity_parts(stream, num_qr, per_each, parity, encoding))]

    return ver, parts

# EOF

File: js/src/split.ts
/**
 * (c) Copyright 2024 by Coinkite Inc. This file is in the public domain.
 *
 * Splitting of data and encoding as BBQr QR codes.
 */

import { ENCODING_BYTE_MOD, ENCODING_SPLIT_MOD, HEADER_LEN, MAX_PARITY_INDEX } from './consts';
import { encodeParity } from './gf256';
import { Encoding, FileType, SplitOptions, SplitResult, Version } from './types';
import {
  base64ToBytes,
  bytesToText,
  encodeBytes,
  fileToBytes,
  hexToBytes,
  intToBase36,
  looksLikePsbt,
  padBlock,
  validateSplitOptions,
  versionToChars,
} from './utils';

function numQRNeeded(version: Version, length: number, encoding: Encoding, reserve = 0) {
  const splitMod = ENCODING_SPLIT_MOD[encoding];

  // reserve: chars to leave free in each QR (parity parts are one group longer)
  const baseCap = versionToChars(version) - HEADER_LEN - reserve;

  // adjust capacity to be a multiple of splitMod
  const adjustedCap = baseCap - (baseCap % splitMod);

  const estimatedCount = Math.ceil(length / adjustedCap);

  if (estimatedCount === 1) {
    // if it fits in one QR, we're done
    return { count: 1, perEach: length };
  }

  // the total capacity of our estimated count
  // all but the last QR need to use adjusted capacity to ensure proper split
  const estimatedCap = (estimatedCount - 1) * adjustedCap + baseCap;

  return {
    count: estimatedCap >= length ? estimatedCount : estimatedCount + 1,
    perEach: adjustedCap,
  };
}

function findBestVersion(length: number, opts: Required<SplitOptions>, reserve = 0) {
  const options: { version: Version; count: number; perEach: number }[] = [];

  for (let version = opts.minVersion; version <= opts.maxVersion; version++) {
    const { count, perEach } = numQRNeeded(version, length, opts.encoding, reserve);

    if (opts.minSplit <= count && count <= opts.maxSplit) {
      options.push({ version, count, perEach });
    }
  }

  if (!options.length) {
    throw new Error('Cannot make it fit');
  }

  // pick smallest number of QR, lowest version
  options.sort((a, b) => a.count - b.count || a.version - b.version);

  return options[0];
}

/**
 * Builds the parity parts (text, without headers) for a series of numQR data parts.
 *
 * - any numQR parts out of the whole series recover the data
 * - each data block is padded with 0x80 then zeros to one symbol group more than a
 *   full block, so a rebuilt block always shows where its data ends
 * - Reed-Solomon over GF(2^8): block i is the value at x=i, see gf256.ts
 */
export function parityParts(
  stream: Uint8Array,
  numQR: number,
  perEach: number,
  parity: number,
  encoding: Encoding
) {
  if (numQR < 2) {
    throw new Error('parity needs 2 or more data parts');
  }

  if (numQR + parity > MAX_PARITY_INDEX + 1) {
    throw new Error('too many parts for parity');
  }

  const mod = ENCODING_BYTE_MOD[encoding];
  const splitMod = ENCODING_SPLIT_MOD[encoding];
  const bData = Math.floor(perEach / splitMod) * mod; // bytes in a full data part
  const bRs = bData + mod;

  const blocks: Uint8Array[] = [];

  for (let offset = 0; offset < stream.length; offset += bData) {
    blocks.push(padBlock(stream.slice(offset, offset + bData), bRs));
  }

  if (blocks.length !== numQR) {
    throw new Error('unexpected number of blocks');
  }

  return encodeParity(blocks, numQR + parity).map((p) => bytesToText(p, encoding));
}

/**
 * Converts the input bytes into a series of QR codes, ensuring that the most efficient QR code
 * version is used.
 *
 * NOTE: When the default 'Z' (Zlib) encoding is selected, it is possible that the actual used encoding
 * will be '2' (Base32) in case Zlib compression does not reduce the size of the output.
 *
 * With `parity` set, that many extra QR codes are appended (indexes after the data parts).
 * Any N parts of such a series, N being the number of data parts, recover the data.
 *
 * @param raw The input bytes to split and encode.
 * @param fileType The file type to use. Refer to BBQr spec.
 * @param opts An optional SplitOptions object.
 *
 * @returns An object containing the version of the QR codes, their string parts, and the actual encoding used.
 */
export function splitQRs(
  raw: Uint8Array,
  fileType: string,
  opts: SplitOptions = {}
): SplitResult {
  if (!/^[A-Z]$/.test(fileType)) {
    throw new Error('fileType must be a single uppercase letter A-Z');
  }

  const validatedOpts = validateSplitOptions(opts);

  const { encoding: actualEncoding, raw: stream } = encodeBytes(raw, validatedOpts.encoding);
  const encoded = bytesToText(stream, actualEncoding);

  let layout = findBestVersion(encoded.length, validatedOpts);
  let parity = validatedOpts.parity;

  if (parity && layout.count >= 2) {
    // parity parts are one symbol group longer than data parts; leave room for
    // that, now that we know the data needs more than one QR anyway
    layout = findBestVersion(encoded.length, validatedOpts, ENCODING_SPLIT_MOD[actualEncoding]);
  } else {
    // a single QR has nothing to lose
    parity = 0;
  }

  const { version, count, perEach } = layout;

  const header = `B$${actualEncoding}${fileType}` + intToBase36(count);

  const parts: string[] = [];

  for (let n = 0, offset = 0; offset < encoded.length; n++, offset += perEach) {
    parts.push(header + intToBase36(n) + encoded.slice(offset, offset + perEach));
  }

  if (parity) {
    parityParts(stream, count, perEach, parity, actualEncoding).forEach((p, n) => {
      parts.push(header + intToBase36(count + n) + p);
    });
  }

  return { version, parts, encoding: actualEncoding };
}

/**
 * Takes a given given input (Uint8Array, File, or string) and detects its FileType.
 * PSBTs and Bitcoin transactions are supported in raw binary, Base64, or hex format.
 *
 * @param input - The input to detect the FileType of.
 * @returns A Promise that resolves to an object containing the FileType and raw data.
 */
export async function detectFileType(
  input: File | Uint8Array | string
): Promise<{ fileType: FileType; raw: Uint8Array }> {
  // keep references to both raw and decoded versions of the input to run checks on
  let raw: Uint8Array | undefined = undefined;
  let decoded: string | undefined = undefined;

  if (input instanceof File) {
    // convert a File to Uint8Array so we have access to the raw bytes
    input = await fileToBytes(input);
  }

  if (input instanceof Uint8Array) {
    // we got binary, see if we recognize it
    raw = input;

    if (looksLikePsbt(input)) {
      console.debug('Detected type "P" from binary input');
      return { fileType: 'P', raw };
    }

    if (raw[0] === 0x01 || raw[0] === 0x02) {
      console.debug('Detected type "T" from binary input');
      return { fileType: 'T', raw };
    }

    // otherwise, try to decode as text (could be contents of a file)
    try {
      decoded = new TextDecoder('utf-8', { fatal: true }).decode(raw);
    } catch (err) {
      // not text, so fall back to generic binary

      console.debug('Detected type "B" from binary input');
      return { fileType: 'B', raw };
    }
  } else if (typeof input === 'string') {
    decoded = input;
  } else {
    throw new Error('Invalid input - must be a File, Uint8Array or string');
  }

  const trimmed = decoded.trim();

  if (/^70736274ff[0-9A-Fa-f]+$/.test(trimmed)) {
    // PSBT in hex format
    console.debug('Detected type "P" from hex input');

    return { fileType: 'P', raw: hexToBytes(trimmed) };
  }

  if (/^0[1,2]000000[0-9A-Fa-f]+$/.test(trimmed)) {
    // Transaction in hex format
    console.debug('Detected type "T" from hex input');

    return { fileType: 'T', raw: hexToBytes(trimmed) };
  }

  if (/^[A-Za-z0-9+/=]+$/.test(trimmed)) {
    // looks like base64 - could be PSBT or transaction
    const bytes = base64ToBytes(decoded);

    if (looksLikePsbt(bytes)) {
      console.debug('Detected type "P" from base64 input');
      return { fileType: 'P', raw: bytes };
    }

    if (bytes[0] === 0x01 || bytes[0] === 0x02) {
      console.debug('Detected type "T" from base64 input');
      return { fileType: 'T', raw: bytes };
    }
  }

  // ensure we have raw bytes for the next step
  raw = raw ?? new TextEncoder().encode(decoded);

  try {
    JSON.parse(decoded);
    console.debug('Detected type "J"');
    return { fileType: 'J', raw };
  } catch (err) {
    // not JSON - fall back to generic Unicode

    console.debug('Detected type "U"');
    return { fileType: 'U', raw };
  }
}

// EOF
